### ISSUE TITLE
Search coarse channel drops one drift block subrange

### ISSUE THREAD

**Issue by texadactyl** [UCBerkeleySETI/turbo_seti#141]
There is one drift block sub-range missing.  The suspect code is in find_doppler.py lines 341-353:
```
        if drift_block <= 0:
            complete_drift_range = data_obj.drift_rate_resolution * fd.kernels.np.array(
                range(-1 * tsteps_valid * (abs(drift_block) + 1) + 1,
                      -1 * tsteps_valid * (abs(drift_block)) + 1))
            sub_range = complete_drift_range[(complete_drift_range < min_drift) &
                                             (complete_drift_range >= -1 * max_drift)]
            logger.debug('drift_block <= 0: sub_range={}'.format(sub_range))
        else:
            complete_drift_range = data_obj.drift_rate_resolution * fd.kernels.np.array(
                range(tsteps_valid * drift_block, tsteps_valid * (drift_block + 1)))
            sub_range = complete_drift_range[(complete_drift_range >= min_drift) &
                                             (complete_drift_range <= max_drift)]
            logger.debug('drift_block > 0: sub_range={}'.format(sub_range))
```

The first time drift_block > 0, the complete_drift_range calculation skips one sub-range.

**Comment by texadactyl** [comment]
The indexes on the range for the following is incorrect in the drift_block > 0 case (lines 351-352):
```
            complete_drift_range = data_obj.drift_rate_resolution * fd.kernels.np.array(
                range(tsteps_valid * drift_block, tsteps_valid * (drift_block + 1)))
```
This block should be:
```
            complete_drift_range = data_obj.drift_rate_resolution * fd.kernels.np.array(
                range(1 + tsteps_valid * (drift_block - 1), 1 + tsteps_valid * (drift_block)))
```

Sample log from issue #135 when this issue is fixed: 
[issue_235.log.zip](https://github.com/UCBerkeleySETI/turbo_seti/files/5739895/issue_235.log.zip)

**Comment by texadactyl** [comment]
The fix to this one created issue #150 !  )-:

**Comment by texadactyl** [comment]
To reproduce, 
* Download https://storage.googleapis.com/ata_test_data/turbo_seti_test/guppi_59196_68468_000762_Unknown_0001.rawspec.0000.h5
* ```turboSETI  guppi_59196_68468_000762_Unknown_0001.rawspec.0000.h5 -n 256 -o /tmp/run_turbo_seti/ -l debug -o /tmp/run_turbo_seti/ | tee capture.log```
* Examine capture.log.  In the first coarse channel (0), find drift_block=0.  Notice the frequency subrange gap between drift_block 0 and 1.

**Comment by texadactyl** [comment]
Subrange details: drift blocks -1, 0, and 1.
```
find_doppler.0  DEBUG    Drift_block -1 (in range from -10 through 10)
find_doppler.0  DEBUG    Un-flipped corrected negative drift_block
find_doppler.0  DEBUG    drift_block <= 0: sub_range=[-1.92509712 -1.90993887 -1.89478063 -1.87962238 -1.86446414 -1.84930589
 -1.83414765 -1.8189894  -1.80383116 -1.78867291 -1.77351467 -1.75835642
 -1.74319818 -1.72803993 -1.71288169 -1.69772344 -1.6825652  -1.66740695
 -1.65224871 -1.63709046 -1.62193222 -1.60677397 -1.59161573 -1.57645748
 -1.56129924 -1.54614099 -1.53098275 -1.5158245  -1.50066626 -1.48550801
 -1.47034977 -1.45519152 -1.44003328 -1.42487503 -1.40971679 -1.39455854
 -1.3794003  -1.36424205 -1.34908381 -1.33392556 -1.31876732 -1.30360907
 -1.28845083 -1.27329258 -1.25813434 -1.24297609 -1.22781785 -1.2126596
 -1.19750136 -1.18234311 -1.16718487 -1.15202662 -1.13686838 -1.12171013
 -1.10655189 -1.09139364 -1.0762354  -1.06107715 -1.04591891 -1.03076066
 -1.01560242 -1.00044417 -0.98528593 -0.97012768]

find_doppler.0  DEBUG    Drift_block 0 (in range from -10 through 10)
find_doppler.0  DEBUG    Un-flipped corrected negative drift_block
find_doppler.0  DEBUG    drift_block <= 0: sub_range=[-0.95496944 -0.93981119 -0.92465295 -0.9094947  -0.89433646 -0.87917821
 -0.86401997 -0.84886172 -0.83370348 -0.81854523 -0.80338699 -0.78822874
 -0.7730705  -0.75791225 -0.74275401 -0.72759576 -0.71243752 -0.69727927
 -0.68212103 -0.66696278 -0.65180454 -0.63664629 -0.62148805 -0.6063298
 -0.59117156 -0.57601331 -0.56085507 -0.54569682 -0.53053858 -0.51538033
 -0.50022209 -0.48506384 -0.4699056  -0.45474735 -0.43958911 -0.42443086
 -0.40927262 -0.39411437 -0.37895613 -0.36379788 -0.34863964 -0.33348139
 -0.31832315 -0.3031649  -0.28800666 -0.27284841 -0.25769017 -0.24253192
 -0.22737368 -0.21221543 -0.19705719 -0.18189894 -0.1667407  -0.15158245
 -0.13642421 -0.12126596 -0.10610772 -0.09094947 -0.07579123 -0.06063298
 -0.04547474 -0.03031649 -0.01515825]

find_doppler.0  DEBUG    Drift_block 1 (in range from -10 through 10)
find_doppler.0  DEBUG    drift_block > 0: sub_range=[0.97012768 0.98528593 1.00044417 1.01560242 1.03076066 1.04591891
 1.06107715 1.0762354  1.09139364 1.10655189 1.12171013 1.13686838
 1.15202662 1.16718487 1.18234311 1.19750136 1.2126596  1.22781785
 1.24297609 1.25813434 1.27329258 1.28845083 1.30360907 1.31876732
 1.33392556 1.34908381 1.36424205 1.3794003  1.39455854 1.40971679
 1.42487503 1.44003328 1.45519152 1.47034977 1.48550801 1.50066626
 1.5158245  1.53098275 1.54614099 1.56129924 1.57645748 1.59161573
 1.60677397 1.62193222 1.63709046 1.65224871 1.66740695 1.6825652
 1.69772344 1.71288169 1.72803993 1.74319818 1.75835642 1.77351467
 1.78867291 1.80383116 1.8189894  1.83414765 1.84930589 1.86446414
 1.87962238 1.89478063 1.90993887 1.92509712]
```
The transition from using the "<= 0" formula to the "else" formula causes a 0.985 gap.

cc: @wfarah @luigifcruz

**Comment by texadactyl** [comment]
If you change the if comparison from "<=" to "<", then drift block 0 joins the "else" computation and the gap moves as expected.
```
find_doppler.0  DEBUG    Drift_block -1 (in range from -10 through 10)
find_doppler.0  DEBUG    Un-flipped corrected negative drift_block
find_doppler.0  DEBUG    drift_block < 0: sub_range=[-1.92509712 -1.90993887 -1.89478063 -1.87962238 -1.86446414 -1.84930589
 -1.83414765 -1.8189894  -1.80383116 -1.78867291 -1.77351467 -1.75835642
 -1.74319818 -1.72803993 -1.71288169 -1.69772344 -1.6825652  -1.66740695
 -1.65224871 -1.63709046 -1.62193222 -1.60677397 -1.59161573 -1.57645748
 -1.56129924 -1.54614099 -1.53098275 -1.5158245  -1.50066626 -1.48550801
 -1.47034977 -1.45519152 -1.44003328 -1.42487503 -1.40971679 -1.39455854
 -1.3794003  -1.36424205 -1.34908381 -1.33392556 -1.31876732 -1.30360907
 -1.28845083 -1.27329258 -1.25813434 -1.24297609 -1.22781785 -1.2126596
 -1.19750136 -1.18234311 -1.16718487 -1.15202662 -1.13686838 -1.12171013
 -1.10655189 -1.09139364 -1.0762354  -1.06107715 -1.04591891 -1.03076066
 -1.01560242 -1.00044417 -0.98528593 -0.97012768]
find_doppler.0  DEBUG    Drift_block 0 (in range from -10 through 10)
find_doppler.0  DEBUG    Un-flipped corrected negative drift_block
find_doppler.0  DEBUG    drift_block >= 0: sub_range=[0.         0.01515825 0.03031649 0.04547474 0.06063298 0.07579123
 0.09094947 0.10610772 0.12126596 0.13642421 0.15158245 0.1667407
 0.18189894 0.19705719 0.21221543 0.22737368 0.24253192 0.25769017
 0.27284841 0.28800666 0.3031649  0.31832315 0.33348139 0.34863964
 0.36379788 0.37895613 0.39411437 0.40927262 0.42443086 0.43958911
 0.45474735 0.4699056  0.48506384 0.50022209 0.51538033 0.53053858
 0.54569682 0.56085507 0.57601331 0.59117156 0.6063298  0.62148805
 0.63664629 0.65180454 0.66696278 0.68212103 0.69727927 0.71243752
 0.72759576 0.74275401 0.75791225 0.7730705  0.78822874 0.80338699
 0.81854523 0.83370348 0.84886172 0.86401997 0.87917821 0.89433646
 0.9094947  0.92465295 0.93981119 0.95496944]
find_doppler.0  DEBUG    Drift_block 1 (in range from -10 through 10)
find_doppler.0  DEBUG    drift_block >= 0: sub_range=[0.97012768 0.98528593 1.00044417 1.01560242 1.03076066 1.04591891
 1.06107715 1.0762354  1.09139364 1.10655189 1.12171013 1.13686838
 1.15202662 1.16718487 1.18234311 1.19750136 1.2126596  1.22781785
 1.24297609 1.25813434 1.27329258 1.28845083 1.30360907 1.31876732
 1.33392556 1.34908381 1.36424205 1.3794003  1.39455854 1.40971679
 1.42487503 1.44003328 1.45519152 1.47034977 1.48550801 1.50066626
 1.5158245  1.53098275 1.54614099 1.56129924 1.57645748 1.59161573
 1.60677397 1.62193222 1.63709046 1.65224871 1.66740695 1.6825652
 1.69772344 1.71288169 1.72803993 1.74319818 1.75835642 1.77351467
 1.78867291 1.80383116 1.8189894  1.83414765 1.84930589 1.86446414
 1.87962238 1.89478063 1.90993887 1.92509712]
```

**Comment by texadactyl** [comment]
Ironically, the symproms of issue #135 are no longer present in either of the above cases (regardless of where the gap is).

**Comment by texadactyl** [comment]
The issue may stem from the original implementation of numba and cupy/gpu support.
Attention is now on find_doppler.py search_coarse_channel.py.

**Comment by texadactyl** [comment]
Fixed in 2.0.7
